R/transformCounts.R

```diff
@@ -308,6 +351,115 @@ setMethod("transformAssay", signature = c(x = "SummarizedExperiment"),
+    
+    # Normalize the count data by dividing by the scaling factor
+    normalized_data <- sweep(mat, 2, scaling_factors / scaling, "/")
+    return(normalized_data)
+}
+
+####################################.calc_css_fast##############################
+# This function applies cumulative sum scaling (CSS) to the abundance table 
+# using the fast variant method.
+.calc_css_fast <- function(mat, css.percentile = NULL, ...) {
+    # Calculate the percentile using .calc_css_percentile
+    if(is.null(css.percentile)) {
+        css.percentile <- .calc_css_percentile(mat, ...)
+    }
+    message("css.percentile value used (", css.percentile, ").")
+    # Call .calc_css with the calculated percentile
+    normalized_data <- .calc_css(mat, css.percentile, ...)
+    return(normalized_data)
+}
+
+#################################.calc_css_percentile######################
+#' @importFrom DelayedMatrixStats colSums2 colQuantiles rowMeans2 rowMedians
+
+# Calculates the cumulative sum scaling (css) scaling percentiles from the given 
+# data
+.calc_css_percentile <- function(mat, rel = 0.1, ...) {
+    # Input check
+    if( !is.numeric(rel) ){
+        stop("'rel' must be a numeric value.", call. = FALSE)
+    }
+    # Replace zero values to NA, i.e. not detected
+    mat[ mat == 0 ] <- NA
+    # Calculate number of features detected
+    found_features <- colSums2(!is.na(mat))
+    # Stop if there are columsn with only 1 or 0 found features
+    if( any(found_features <= 1) ){
+        stop(
+            "There are samples that contain 1 or less found features.",
+            "'fast CSS' method cannot be applied.", call. = FALSE)
+    }
+    # Calculate quantiles. Take into account only positive, found features
+    quantiles <- colQuantiles(
+        mat, probs = seq(0, 1, length.out = max(found_features)), na.rm = TRUE)
+    quantiles <- t(quantiles)
+    # Sort the columns based on abundance
+    mat <- apply(mat, 2, function(col){
+        col[ is.na(col) ] <- 0
+        col <- as.array(col)
+        col <- sort(col)
+        return(col)
+    })
+    # Compute the row means of sorted matrix to create a reference profile
+    ref <- rowMeans2(mat)
+    ref <- ref[ ref > 0 ]
+    # Calculate the differences between the reference profile and each
+    # column's quantiles
+    difference <- ref - quantiles
+    # Compute the row medians of absolute differences
+    difference <- rowMedians(abs(difference))
+    # Find the first point where the relative change exceeds the threshold
+    res <- abs(diff(difference)) / difference[-1]
+    res <- which(res > rel)
+    res <- res[[1]] / max(found_features)
+    # If the calculated percentile is less than or equal to 0.50, use 0.50 as
+    # the default value
+    if( res < 0.50 ){
+        res <- 0.50
+        message("Default rel value used (", res, ").")
+    } else {
+        message("rel value used (", res, ").")
+    }
+    return(res)
```

Comment:
From where this 0.5 comes from?

Reply:
it is referenced from the percentile

Reply:
The documentation says that the default value of percentile is 0.5. However, that is not true as it is calculated (it is >= 0.5)

I was also wondering what makes the 0,5 special number. Why it cannot be lower by default. However, I think that is ok since that is calculated only if user does not specify the value